### PR TITLE
Implement new REST API to retrieve API artifacts for a list of API UUIDs

### PR DESCRIPTION
**Pull request by Jayanie**
## Purpose

- This fixes https://github.com/wso2-enterprise/choreo/issues/5941
- Introduces a New REST API as `/retrieve-api-artifacts` with a post resource to pass the list of the API UUIDs as a payload in the request and obtain the API artifacts zip.

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.internal.service/src/main/java/org/wso2/carbon/apimgt/internal/service/impl/RetrieveApiArtifactsApiServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
+package org.wso2.carbon.apimgt.internal.service.impl;
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.internal.service/src/main/java/org/wso2/carbon/apimgt/internal/service/impl/RetrieveApiArtifactsApiServiceImpl.java:1]: Missing license.

File: components/apimgt/org.wso2.carbon.apimgt.internal.service/src/main/resources/api.yaml [hunk 2]
```diff
@@ -476,6 +476,39 @@ paths:
...
+        - name: gatewayLabel
+          in: query
+          description: |
+            **Search condition**.
+             label associated with the APIs
+          type: string
+        - name: type
+          in: query
+          description: |
+            **Search condition**.
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.internal.service/src/main/resources/api.yaml:495]: Is this part required
```suggestion
```

File: components/apimgt/org.wso2.carbon.apimgt.internal.service/src/main/resources/api.yaml [hunk 3]
```diff
@@ -476,6 +476,39 @@ paths:
...
+             type of gateway
+          type: string
+        - name: uuidList
+          in: body
+          description: 'API UUID list payload'
+          schema:
+            $ref: "#/definitions/UUIDList"
+      tags:
+        - Retrieving API artifacts
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.internal.service/src/main/resources/api.yaml:504]:
```suggestion
        - Retrieve API artifacts
```

File: components/apimgt/org.wso2.carbon.apimgt.internal.service/src/main/resources/api.yaml [hunk 4]
```diff
@@ -476,6 +476,39 @@ paths:
           description: Unexpected error
           schema:
             $ref: '#/definitions/Error'
+  /retrieve-api-artifacts:
+    post:
+      summary: Get all the API runtime artifacts for a given API UUID list and gateway label
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.internal.service/src/main/resources/api.yaml:481]: Summary is too long. Let's make it shorter and add more information to the description
```suggestion
      summary: Get API runtime artifacts from ID list
```

File: components/apimgt/org.wso2.carbon.apimgt.internal.service/src/main/java/org/wso2/carbon/apimgt/internal/service/impl/RetrieveApiArtifactsApiServiceImpl.java (hunk 1 above)
Comment (Jayanie) [reply components/apimgt/org.wso2.carbon.apimgt.internal.service/src/main/java/org/wso2/carbon/apimgt/internal/service/impl/RetrieveApiArtifactsApiServiceImpl.java:18]: Fixed with f86e74c9ac369f1a34db199d0c4f518c48f40737

File: components/apimgt/org.wso2.carbon.apimgt.internal.service/src/main/resources/api.yaml (hunk 2 above)
Comment (Jayanie) [reply components/apimgt/org.wso2.carbon.apimgt.internal.service/src/main/resources/api.yaml]: Fixed with 87f85e02881b3d64d82f300d60ee009702fb35d5

File: components/apimgt/org.wso2.carbon.apimgt.internal.service/src/main/resources/api.yaml (hunk 3 above)
Comment (Jayanie) [reply components/apimgt/org.wso2.carbon.apimgt.internal.service/src/main/resources/api.yaml]: Fixed with 87f85e02881b3d64d82f300d60ee009702fb35d5

File: components/apimgt/org.wso2.carbon.apimgt.internal.service/src/main/resources/api.yaml (hunk 4 above)
Comment (Jayanie) [reply components/apimgt/org.wso2.carbon.apimgt.internal.service/src/main/resources/api.yaml]: Fixed with 87f85e02881b3d64d82f300d60ee009702fb35d5

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/GatewayArtifactsMgtDAO.java [hunk 5]
```diff
@@ -358,6 +358,63 @@ public void deleteGatewayArtifacts(String apiId) throws APIManagementException {
         return apiRuntimeArtifactDtoList;
     }
 
+    public List<APIRuntimeArtifactDto> retrieveGatewayArtifactsByAPIIDs(List<String> apiIds, String[] labels,
+                                                                        String tenantDomain)
+            throws APIManagementException {
+
+        String query = SQLConstants.RETRIEVE_ARTIFACTS_BY_APIID_AND_LABEL;
+        query = query.replaceAll("_GATEWAY_LABELS_", String.join(",",Collections.nCopies(labels.length, "?")));
```
[the remark this answers is not in the document]
Comment (Jayanie) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/GatewayArtifactsMgtDAO.java]: Fixed with 927627e64e1e4ebd6c471a39ccbf821c6248bda0

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/GatewayArtifactsMgtDAO.java [hunk 6]
```diff
@@ -358,6 +358,63 @@ public void deleteGatewayArtifacts(String apiId) throws APIManagementException {
...
+                    index++;
+                }
+                preparedStatement.setString(index, tenantDomain);
+                try (ResultSet resultSet = preparedStatement.executeQuery()) {
+                    while (resultSet.next()) {
+                        APIRuntimeArtifactDto apiRuntimeArtifactDto = new APIRuntimeArtifactDto();
+                        apiRuntimeArtifactDto.setTenantDomain(resultSet.getString("TENANT_DOMAIN"));
+                        apiRuntimeArtifactDto.setApiId(apiId);
+                        String label = resultSet.getString("LABEL");
+                        // Do not handle the exception here since runtime artifacts are retrieved by API UUID
```
[the remark this answers is not in the document]
Comment (Jayanie) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/GatewayArtifactsMgtDAO.java]: Fixed with 927627e64e1e4ebd6c471a39ccbf821c6248bda0

**Review by malinthaprasan — APPROVED** [review]

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/GatewayArtifactsMgtDAO.java [hunk 7]
```diff
@@ -358,11 +358,66 @@ public void deleteGatewayArtifacts(String apiId) throws APIManagementException {
...
+                                                                        String tenantDomain)
+            throws APIManagementException {
+
+        String query = SQLConstants.RETRIEVE_ARTIFACTS_BY_APIID_AND_LABEL;
+        query = query.replaceAll(SQLConstants.GATEWAY_LABEL_REGEX, String.join(",",Collections.nCopies(labels.length, "?")));
+        List<APIRuntimeArtifactDto> apiRuntimeArtifactDtoList = new ArrayList<>();
+        try (Connection connection = GatewayArtifactsMgtDBUtil.getArtifactSynchronizerConnection();
+             PreparedStatement preparedStatement = connection.prepareStatement(query)) {
+            for (String apiId: apiIds) {
+                preparedStatement.setString(1, apiId);
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/GatewayArtifactsMgtDAO.java:371]: If I understand correctly, from this, we are sending one query per each APIID?
Here the number of APIIDs can be high like 100, 1000 2000 etc right.. we are looking into those numbers right? In that case its also bit dangerous to use IN with all API IDs at once.. 
Can we split the List<String> apiIds into smaller lists (may be one containing 25 APIIDs), then when processing each smaller list we can do an IN altogether? Then we reduce the query count in 25 times..

Comment (Jayanie) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/GatewayArtifactsMgtDAO.java]: Fixed with 08e485e24c0c161e5fe355208c8cde2a7ceba5df

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/GatewayArtifactsMgtDAO.java [hunk 8]
```diff
@@ -358,11 +355,86 @@ public void deleteGatewayArtifacts(String apiId) throws APIManagementException {
...
+        // Split apiId list into smaller list of size 25
+        List<List<String>> apiIdsChunk = new ArrayList<>();
+        int apiIdListSize = apiIds.size();
+        int apiIdArrayIndex = 0;
+        int apiIdsChunkSize = 25;
+        while (apiIdsChunkSize < apiIdListSize) {
+            apiIdsChunk.add(apiIds.subList(apiIdArrayIndex, apiIdsChunkSize));
+            apiIdListSize = apiIdListSize - 25;
+            apiIdArrayIndex = apiIdArrayIndex + 25;
+            apiIdsChunkSize = apiIdsChunkSize + 25;
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/GatewayArtifactsMgtDAO.java:370]: Use a constant for 25

**Review by malinthaprasan — APPROVED** [review]

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/GatewayArtifactsMgtDAO.java (hunk 8 above)
Comment (Jayanie) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/GatewayArtifactsMgtDAO.java]: Fixed with bf3bd02aac15a4788edd4d35e26bf1d5208a0471

**Review by mpmunasinghe — APPROVED** [review]
LGTM

**Pull request merged**